src/names/scopes.md

```diff
@@ -1,3 +1,329 @@
+[Methods] can also be referred to via [call expressions].
+
+Similar to items within a module or block, it is an error to introduce an item within a trait or implementation that is a duplicate of another item in the trait or impl in the same namespace.
+
+## Pattern binding scopes
+
+The scope of a local variable [pattern] binding depends on where it is used:
+
+* [`let` statement] bindings range from just after the `let` statement until the end of the block where it is declared.
+* [Function parameter] bindings are within the body of the function.
+* [Closure parameter] bindings are within the closure body.
+* [`for`] and [`while let`] bindings are within the loop body.
+* [`if let`] bindings are within the consequent block.
+* [`match` arms] bindings are within the [match guard] and the match arm expression.
+
+Local variable scopes do not extend into item declarations.
+<!-- Not entirely, see https://github.com/rust-lang/rust/issues/33118 -->
+
+### Pattern binding shadowing
+
+Pattern bindings are allowed to shadow any name in scope with the following exceptions which are an error:
+
+* [Const generic parameters]
+* [Static items]
+* [Const items]
+* Constructors for [structs] and [enums]
+
+The following example illustrates how local bindings can shadow item declarations:
+
+```rust
+fn shadow_example() {
+    // Since there are no local variables in scope yet, this resolves to the function.
+    foo(); // prints `function`
+    let foo = || println!("closure");
+    fn foo() { println!("function"); }
+    // This resolves to the local closure since it shadows the item.
+    foo(); // prints `closure`
+}
+```
+
+## Generic parameter scopes
+
+Generic parameters are declared in a [_GenericParams_] list.
+The scope of a generic parameter is within the item it is declared on.
+
+All parameters are in scope within the generic parameter list regardless of the order they are declared.
+The following shows some examples where a parameter may be referenced before it is declared:
+
+```rust
+// The 'b bound is referenced before it is declared.
+fn params_scope<'a: 'b, 'b>() {}
+
+# trait SomeTrait<const Z: usize> {}
+// The const N is referenced in the trait bound before it is declared.
+fn f<T: SomeTrait<N>, const N: usize>() {}
+```
+
+Generic parameters are also in scope for type bounds and where clauses, for example:
+
+```rust
+# trait SomeTrait<'a, T> {}
+// The <'a, U> for `SomeTrait` refer to the 'a and U parameters of `bounds_scope`.
+fn bounds_scope<'a, T: SomeTrait<'a, U>, U>() {}
+
+fn where_scope<'a, T, U>()
+    where T: SomeTrait<'a, U>
+{}
+```
+
+Generic scopes do not extend into [items] declared inside a function.
+
+### Generic parameter shadowing
+
+It is an error to shadow a generic parameter.
+Items declared within functions are allowed to reuse generic parameter names from the function because generic scopes do not extend to inner items.
```

Comment:
The motivation after "because" is artificial.
This is allowed because shadowing of names is allowed in general, items in inner scopes can shadow items from outer scopes, et cetera.

The restrictions on generic parameters that appear in some cases are exceptions.
They exist for historical reasons, but technically nothing prevents removing them at any moment.

Reply:
I reworded this a bit.